### ISSUE TITLE
Make an Ast Diff tool for javaparser?

### ISSUE THREAD

**Issue by arturbosch** [javaparser/javaparser#710]
Do you know about a project or prototype which support diffs? My usecase is to track the evolution of some code fragments. Before I start a prototype in that direction I just wanted to be sure.

**Comment by ftomassetti** [comment]
I am not aware of such a tool but this is **definitely** something I would interest into.

**Comment by matozoid** [comment]
@ftomassetti - is your observer functionality not useful for this?

**Comment by ftomassetti** [comment]
no, I would say no. I think this tool would be about getting two commits or two checkouts and compare the two versions of each file. You would get two ASTs and you would need to find the correspondances. So, you could use the name in many cases (classes, methods, etc) but sometimes you should use the content (for example to realize a method was renamed but it kept the same body). Finally you could derive a list of changes.

The observer could be useful to create a list of changes if someone did the change directly on the AST. In that case yes, the observer would be perfect.

**Comment by arturbosch** [comment]
yes, @ftomassetti nailed it. Another use case would be to detect clone fragments with slight changes. For example you have two methods but only the method and variable names are different.

**Comment by matozoid** [comment]
Whew, that's quite an interesting thing to build.

**Comment by ptitjes** [comment]
I think what you are looking for is a **semantic diff** tool.
Websearch https://duckduckgo.com/?q=java+semantic+diff

**Comment by ftomassetti** [comment]
well, yes, ideally yes, but initially we could start with a simple syntactic comparison based on the AST, and then make it smarter and move to a more semantic diff

**Comment by shashanksingh28** [comment]
something like [GumTree](https://github.com/GumTreeDiff/gumtree/)? I just found this out while I was searching for "graph edit distance for code AST's"

**Comment by matozoid** [comment]
Relabeling as "fun project idea" because we don't have the manpower to support more features.

**Comment by Symbolk** [comment]
For AST diff, we are currently using the **SpoonLabs/gumtree-spoon-ast-diff,** which uses the INRIA/Spoon as the Java code parser but contains several bugs.

**Comment by verhasi** [comment]
Hi @ftomassetti, @matozoid, @jlerbsc  and all!

I have crafted a tool https://bitbucket.org/mocker-guru/semantic-java-comparator/src/master/ that could be called "simple syntactic comparison based on the AST, and then make it smarter and move to a more semantic diff" if you are still interested in after about a decade. Any feedback is welcomed.
